crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -121,15 +123,47 @@ impl Plugin for VisibilityPlugin {
 pub fn calculate_bounds(
     mut commands: Commands,
     meshes: Res<Assets<Mesh>>,
-    without_aabb: Query<(Entity, &Handle<Mesh>), (Without<Aabb>, Without<NoFrustumCulling>)>,
+    without_aabb_or_with_changed_mesh: Query<
+        (Entity, &Handle<Mesh>),
+        (
+            Or<(Without<Aabb>, Changed<Handle<Mesh>>)>,
+            Without<NoFrustumCulling>,
+        ),
+    >,
+    mut entity_mesh_map: Local<HashMap<Handle<Mesh>, Vec<Entity>>>,
+    mut mesh_events: EventReader<AssetEvent<Mesh>>,
 ) {
-    for (entity, mesh_handle) in without_aabb.iter() {
+    for (entity, mesh_handle) in without_aabb_or_with_changed_mesh.iter() {
+        // Record entities that have mesh handles.
+        entity_mesh_map
+            .entry(mesh_handle.clone())
+            .or_default()
+            .push(entity);
+
         if let Some(mesh) = meshes.get(mesh_handle) {
             if let Some(aabb) = mesh.compute_aabb() {
                 commands.entity(entity).insert(aabb);
             }
         }
     }
+
+    // Calculate bounds for entities whose meshes have been mutated.
+    let updated_mesh_handles = mesh_events.iter().filter_map(|event| match event {
+        AssetEvent::Modified { handle } => Some(handle),
+        _ => None,
+    });
+    let updated_meshes_and_entities = updated_mesh_handles.filter_map(|mesh_handle| {
+        meshes
+            .get(mesh_handle)
+            .zip(entity_mesh_map.get(mesh_handle))
+    });
+    for (mesh, entities) in updated_meshes_and_entities {
+        if let Some(aabb) = mesh.compute_aabb() {
+            for entity in entities {
+                commands.entity(*entity).insert(aabb.clone());
+            }
+        }
+    }
```

Comment:
`zip` is fine here, but I'd suggest the following code.
```rust
let to_update = |event| {
    let handle = match event {
        AssetEvent::Modified { handle } => handle,
        _ => return None,   
     };
    let mesh = meshes.get(handle)?;
    let with_handle = entity_mesh_map.get(handle)?;
    let aabb = mesh.compute_aabb()?;
    Some((aabb, with_handle))
};
for (aabb, with_handle) in events.iter().filter_map(to_update) {
    for to_update in with_handle {
        commands.entity(to_update).insert(aabb.clone());
    }
}```

Reply:
Awesome! Updated in 079a9930600907484085b0b2c1631f4a4b54751b.